### PR TITLE
Ensure that a Rewards wallet has been created for iOS QR code modal

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/12872

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [x] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Open browser with a clean profile.
- Navigate to `brave://rewards`.
- On the onboarding modal, click the "X" button to dismiss.
- Click the "QR Code" button.
  - Verify that a QR code image appears.

### CODE REVIEW

**Review by emerick — APPROVED** [review]
LGTM

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 1]
```diff
@@ -1892,6 +1894,15 @@ void RewardsDOMHandler::GetPaymentId(const base::ListValue* args) {
...
 
+  // Ensure that a wallet has been created for the user before attempting
+  // to retrieve a payment ID.
+  rewards_service_->CreateWallet(
+      base::BindOnce(&RewardsDOMHandler::OnWalletCreatedForPaymentId,
+          weak_factory_.GetWeakPtr()));
+}
+
+void RewardsDOMHandler::OnWalletCreatedForPaymentId(
+    ledger::type::Result result) {
```
Comment (marshall) [inline browser/ui/webui/brave_rewards_page_ui.cc:1905]: should there be any logic validating the value of `result`?

also, what happens when `GetPaymentId` is called and a wallet already exists?

**Review by marshall — COMMENTED** [review]
the change looks simple, but can you follow up about the safety of ignoring `result` in the `CreateWallet` handler? assuming you have those figured out I'm good with it

File: browser/ui/webui/brave_rewards_page_ui.cc (hunk 1 above)
Comment (zenparsing) [reply browser/ui/webui/brave_rewards_page_ui.cc:1905]: If there's an error creating the wallet, then `GetBraveWallet` will eventually return an empty wallet pointer, which will get sent to JS as an empty-string payment ID.

If the wallet already exists, then `rewards_service_->CreateWallet(...)` will be a no-op, and `GetBraveWallet` will get the  (already existing) payment ID.

**Review by marshall — APPROVED** [review]
sounds good based on explanation by @zenparsing

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 10 x64` using the following build:

```
Brave | 1.19.43 Chromium: 87.0.4280.67 (Official Build) nightly (64-bit)
-- | --
Revision | 0e5d92df40086cf0050c00f87b11da1b14580930-refs/branch-heads/4280@{#1441}
OS | Windows 10 OS Version 2009 (Build 19042.630)
```

**`Case 1 - Skipping/dismissing "Earn Tokens & Give Back" modal`**

* ensured that `brave://rewards-internals` is empty on first launch when  `Earn Tokens & Give Back` modal is dismissed
* ensured that the QR code was created and visible once clicking on `QR code` after dismissing the `Earn Tokens & Give Back` modal
* ensured that `wallet.payment_id` gets created via `brave://rewards-internals` once `QR code` is pressed
* ensured the `paymentId` persists after several restarts while `ads` & `a-c` were disabled
* ensured that the `paymentId` retreived from the QR code matches `brave://rewards-internals`
* ensured that the same `paymentId` is being used when enabling `ads` & `a-c` via the toggles under `brave://rewards`
* ensured the `paymentId` persists after several restarts after both `ads` & `a-c` were enabled via `brave://rewards`

**`Case 2 - Clicking on "Earn Tokens & Give Back" modal`**

* ensured that the QR code is being created when clicking on `Earn Tokens & Give Back` and enabling `ads` & `a-c`
* ensured that the `paymentId` from the QR code matches `wallet.payment_id` via `brave://rewards-internals`
* ensured the above worked using the following entry points: `rewards panel`, `brave://rewards` and `brave://newtab`

**`Case 3 - Restoring Wallets`**

* ensured that the QR code displayed the correct `paymentId` when a wallet has been restored
* ensured the `paymentId` persists after several restarts after restoring

**Pull request merged**